### PR TITLE
New Rule: UnnecessarySafeCall

### PR DESCRIPTION
**Pull request by cortinico**
Adding the `UnnecessarySafeCall` rule, as suggested in #2572

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        if (bindingContext == BindingContext.EMPTY) return
+
+        val safeAccessElement = expression.getChildOfType<LeafPsiElement>()
+        if (safeAccessElement == null || safeAccessElement.elementType != KtTokens.SAFE_ACCESS) {
+            return
+        }
+
+        if (bindingContext.diagnostics.forElement(safeAccessElement)
+                .any { it.factory == Errors.UNNECESSARY_SAFE_CALL }
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt:54]: For better readability this code could be extracted in a local var.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCallSpec.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                }"""
+            val findings = subject.compileAndLintWithContext(wrapper.env, code)
+            assertThat(findings).hasSize(2)
+            assertThat(findings).hasTextLocations(48 to 59, 48 to 70)
+        }
+    }
+
+    describe("check valid safe operators usage") {
+
+        it("does not reports a simple safe operator usage on nullable type") {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCallSpec.kt:55]:
```suggestion
        it("does not report a simple safe operator usage on nullable type") {
```

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import org.jetbrains.kotlin.resolve.BindingContext
+
+/**
+ * Reports Unnecessary Safe call operators that can be removed by the user.
+ *
+ * <noncompliant>
+ * fun foo(s: String) {
+ *     val a: String = ""
+ *     val b = someValue?.length
+ * }
```
[the remark this answers is not in the document]
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt:24]: The declared function doesn't provide any value to the reader here.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt (hunk 1 above)
Comment (cortinico) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt]: 👍

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * val a: String? = null
+ * val b = someValue?.length
+ * </compliant>
+ */
+class UnnecessarySafeCall(config: Config = Config.empty) : Rule(config) {
+
+    override val issue: Issue = Issue(
+        "UnnecessarySafeCall",
+        Severity.Defect,
+        "Unnecessary Safe call operators detected.",
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt:35]:
```suggestion
        "Unnecessary safe call operator detected.",
```

**Pull request merged**